File: tumour_analysis/analyze_tumour.R
# Define paths
data_path <- "data/sim_av_tumour.csv"
results_path <- "results/"

# Load data
print(paste("Loading data from", data_path))
df <- read.csv(data_path)

# --- Helper Functions & Definitions ---

# ICD-10 Common Cancer Codes Mapping
# Source: Standard ICD-10 definitions
icd10_map <- c(
  "C50" = "Breast",
  "C61" = "Prostate",
  "C34" = "Lung",
  "C18" = "Colon",
  "C19" = "Rectosigmoid",
  "C20" = "Rectum",
  "C43" = "Melanoma",
  "C44" = "Other Skin",
  "C67" = "Bladder",
  "C25" = "Pancreas",
  "C64" = "Kidney",
  "C90" = "Mult. Myeloma",
  "C92" = "Myeloid Leukemia",
  "C54" = "Uterus",
  "D05" = "Carcinoma in situ (Breast)"
)

get_cancer_label <- function(code) {
  if (code %in% names(icd10_map)) {
    return(paste0(code, " (", icd10_map[code], ")"))
  } else {
    return(code)
  }
}

# 1. Top 10 Cancer Sites
if("SITE_ICD10_O2" %in% names(df)) {
  print("Top 10 sites plot saved.")
} else {
  print("SITE_ICD10_O2 column not found.")
}

# 2. Stage Distribution
stage_col <- NULL
if("STAGE_BEST" %in% names(df)) stage_col <- "STAGE_BEST"
if(is.null(stage_col) && "STAGE_BEST_SYSTEM" %in% names(df)) stage_col <- "STAGE_BEST_SYSTEM"

if(!is.null(stage_col)) {
  print(paste("Using stage column:", stage_col))
  
  # Get counts
  stage_counts <- table(df[[stage_col]])
  
  # Filter for top N stages to avoid clutter if there are too many rare codes
  # Or just plot all if reasonable. Let's check the top 15.
  stage_counts_sorted <- sort(stage_counts, decreasing = TRUE)
  top_stages <- head(stage_counts_sorted, 15)
  
  # Scale counts to 10^5
  counts_scaled <- top_stages / 100000
  
  print("Top Stage Counts:")
  print(top_stages)
  
  # Add a legend/explanation text to the plot
  # Explanation of codes (General interpretation)
  # 1, 2, 3, 4: Standard TNM Stages I-IV
  # A, B, C: Sub-stages (e.g., 1A, 1B)
  # ?: Unknown / Not Staged
  # U: Unstaged
  legend_text <- c(
    "1, 2, 3, 4: Main Stages (I-IV)",
    "A, B, C: Sub-stages",
    "?: Unknown / Not Staged",
    "U: Unstaged"
  )
  
  png(paste0(results_path, "stage_distribution.png"), width=900, height=600)
  par(mar=c(5, 5, 4, 2) + 0.1)
  
  # Create barplot
  bp <- barplot(counts_scaled, 
                main = "Distribution of Cancer Stages",
                ylab = "Count (x 10^5)",
                col = "darkorange",
                las = 2, # Rotate x-axis labels
                ylim = c(0, max(counts_scaled) * 1.2)) # Add space for legend
  
  # Add legend
  legend("topright", 
         legend = legend_text, 
         bty = "n", 
         cex = 0.8,
         title = "Code Key")
         
  dev.off()
  print("Stage distribution plot saved.")
} else {
  print("No suitable STAGE column found.")
}

print("Analysis complete.")

File: tumour_analysis/analyze_tumour.py
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import re

# Define paths
data_path = "data/sim_av_tumour.csv"
results_path = "results/"

print(f"Loading data from {data_path}")
try:
    df = pd.read_csv(data_path)
except FileNotFoundError:
    print("File not found.")
    exit()

# --- 1. Top 10 Invasive Cancer Sites ---
print("Analyzing Cancer Sites...")

# Extract 3-char code
df['SITE_MAIN'] = df['SITE_ICD10_O2'].str[:3]

# Filter 1: Keep only Invasive (Starts with 'C')
# "The Fix: Filter your dataset to include only codes starting with 'C'"
df_invasive = df[df['SITE_MAIN'].str.startswith('C', na=False)].copy()

# Filter 2: Exclude C44 (NMSC)
# "The Fix: Filter out C44 before plotting."
df_final = df_invasive[df_invasive['SITE_MAIN'] != 'C44'].copy()

# Count
site_counts = df_final['SITE_MAIN'].value_counts().head(10)

# Map codes to names (Simplified map for demo)
icd10_map = {
    "C50": "Breast", "C61": "Prostate", "C34": "Lung",
    "C18": "Colon", "C19": "Rectosigmoid", "C20": "Rectum",
    "C43": "Melanoma", "C67": "Bladder", "C25": "Pancreas",
    "C64": "Kidney", "C90": "Mult. Myeloma", "C92": "Myeloid Leukemia",
    "C54": "Uterus"
}
labels = [f"{code} ({icd10_map.get(code, code)})" for code in site_counts.index]

# Scale to Thousands (Consistency with previous fixes)
site_counts_scaled = site_counts / 1000

plt.figure(figsize=(10, 6))
sns.barplot(x=site_counts_scaled.values, y=labels, color="steelblue")
plt.title("Top 10 Invasive Cancer Sites (Excl. NMSC)", fontsize=14)
plt.xlabel("Count (Thousands)", fontsize=12)
plt.ylabel("ICD-10 Site", fontsize=12)
plt.tight_layout()
plt.savefig(f"{results_path}top_10_cancer_sites.png")
print("Top 10 plot saved.")


# --- 2. Distribution of Cancer Stages ---
print("Analyzing Stages...")

# Helper to clean stage
def clean_stage(stage):
    stage = str(stage).upper().strip()
    if stage in ['?', 'U', 'NAN', 'nan']:
        return 'Unknown'
    # Aggregate 1A, 1B -> 1
    # Regex to grab the first digit if it starts with 1, 2, 3, 4
    match = re.match(r'^([1-4])', stage)
    if match:
        return match.group(1)
    return 'Unknown' # Treat anything else as unknown for this high-level chart

# Apply cleaning
# Check which column has stage. Usually 'QUIP_T_STAGE' or similar in Simulacrum, 
# but let's check columns or assume a standard name if we can't see them.
# Based on previous R script, it might be 'T_BEST', 'STAGE_BEST', or similar.
# Let's look for a likely column.
stage_cols = [c for c in df.columns if 'STAGE' in c]
if stage_cols:
    target_col = stage_cols[0] # Pick the first one found, likely 'STAGE_BEST' or 'T_BEST'
    print(f"Using column '{target_col}' for staging.")
    
    df['STAGE_CLEAN'] = df[target_col].apply(clean_stage)
    
    # Define Order: 1, 2, 3, 4, Unknown
    stage_order = ['1', '2', '3', '4', 'Unknown']
    
    # Count
    stage_counts = df['STAGE_CLEAN'].value_counts()
    
    # Reindex to enforce order (fill missing with 0)
    stage_counts = stage_counts.reindex(stage_order, fill_value=0)
    
    # Scale
    stage_counts_scaled = stage_counts / 1000
    
    plt.figure(figsize=(8, 6))
    sns.barplot(x=stage_counts_scaled.index, y=stage_counts_scaled.values, color="seagreen")
    
    plt.title("Distribution of Cancer Stages", fontsize=14)
    plt.xlabel("Stage", fontsize=12)
    plt.ylabel("Count (Thousands)", fontsize=12)
    plt.tight_layout()
    plt.savefig(f"{results_path}stage_distribution.png")
    print("Stage plot saved.")
    
else:
    print("No STAGE column found. Skipping stage plot.")
